src/mono/mono/metadata/appdomain.c

```diff
@@ -314,16 +314,17 @@ mono_runtime_init_checked (MonoDomain *domain, MonoThreadStartCB start_cb, MonoT
 	mono_marshal_init ();
 	mono_gc_init_icalls ();
 
-	mono_install_assembly_preload_hook_v2 (mono_domain_assembly_preload, GUINT_TO_POINTER (FALSE), FALSE);
-	mono_install_assembly_search_hook_v2 (mono_domain_assembly_search, GUINT_TO_POINTER (FALSE), FALSE, FALSE);
-	mono_install_assembly_search_hook_v2 (mono_domain_assembly_postload_search, GUINT_TO_POINTER (FALSE), FALSE, TRUE);
-	mono_install_assembly_load_hook_v2 (mono_domain_fire_assembly_load, NULL);
+	// We have to append here because otherwise this will run before the netcore hook (which is installed first), see https://github.com/dotnet/runtime/issues/34273
+	mono_install_assembly_preload_hook_v2 (mono_domain_assembly_preload, GUINT_TO_POINTER (FALSE), FALSE, TRUE);
```

Comment:
Wouldn't it be possible to merge some of the hooks instead ?

Reply:
We'd have to split them, not merge them. The problem is that we have two hooks installed but care about the order they're run in. I don't think we have much to gain from refactoring a bunch of this, but if you feel differently I can explore it a bit more.

Reply:
Keep in mind that even if we change the internal structure of the hooks, we have to keep the v1 embedding APIs still functional (and working alongside them) and port the legacy and netcore loaders to use the new setup. It's not trivial.

Reply:
Added a comment to the issue to clarify the actual bug, sorry for not doing that earlier.